doc/ug/rust_for_c.md

```diff
@@ -0,0 +1,2305 @@
+    In particular, the `()` pattern matches the unit value `()`.
+```rust
+let x: (u32, u32) = ...;
+match x {
+  (5, u) => { ... },  // Check that first element is five, 
+                      // bind the second element to `u`.
+  (u, _) => { ... },  // Bind the first element to `u`,
+                      // discard the second element.
+}
+
+let y: (u32, (u32, u32)) = ...;
+match y {
+  // All patterns can nest arbitrarily, like expressions.
+  (42, (u, _)) =>  { ... },
+  // .. can be used to match either a head or a tail of tuple.
+  (.., u) => { ... },
+  (u, ..) => { ... },
+  (..) => { ... },  // Synonymous with _.
+}
+```
+*   Struct patterns are analogous to tuple patterns.
+    For a tuple-like struct, they have the exact same syntax, but start with the name of the struct: `MyTuple(a, b, _)`.
+    Regular structs are much more interesting syntax-wise:
+```rust
+struct MyStruct { a: i32, b: u32 }
+match my_struct {
+  MyStruct { a, b } => { ... },  // Bind the fields `a` and `b` to
+                                 // names `a` and `b`, respectively.
+  MyStruct { a: foo, b: _ } => { ... }  // Bind the field `a` to the name
+                                   `    // foo`, and discard the field `b`.
+  MyStruct { a: -5, .. } => { ... }  // Check that `a` is -5, and ignore
+                                     // other fields.
+}
+```
+*   Enum patterns are probably the most important kind of pattern, and are what we saw in the `match` statement with `Option` above.
+    They're very similar to struct patterns, except that instead of always succeeding, they check that the enum discriminant is the one named in the pattern.
+```rust
+enum MyEnum { A, B{u32), C { a: i32, b: i32 }, }
+match my_enum {
+  MyEnum::A => { ... }  // Match for variant `A`.
+  MyEnum::B(7) => { ... }  // Match for variant `B`, with 7 as the value inside.
+  MyEnum::B(x) => { ... }  // Match for variant `B`, binding the value inside to
+                           // `x`.
+  MyEnum::C { a: 7, .. } => { ... }  // Match for variant `C`, with 7 as the
+                                     // value in `a` and all other fields ignored.
+  MyEnum::C { b, .. } => { ... }  // Match for variant `C`, binding b to b.
+}
+```
+A complete treatment of the pattern syntax can be found at [https://doc.rust-lang.org/book/ch18-03-pattern-syntax.html](https://doc.rust-lang.org/book/ch18-03-pattern-syntax.html).
+
+A `match` expression will evaluate each pattern against a value until one matches, in order; the compiler will warn about patterns which are unreachable[^87].
+The compiler will also ensure that every value will get matched with one of the match arms, either because every case is covered (e.g., every enum variant is present) or an _irrefutable pattern_ is present (i.e., a patte which matches all values). `_`, `foo`, `(a, _)`, and `MyStruct { a, .. }` are all examples of irrefutable patterns.
```

Comment:
```suggestion
The compiler will also ensure that every value will get matched with one of the match arms, either because every case is covered (e.g., every enum variant is present) or an _irrefutable pattern_ is present (i.e., a pattern which matches all values). `_`, `foo`, `(a, _)`, and `MyStruct { a, .. }` are all examples of irrefutable patterns.
```

Reply:
Done.